x/hold/keeper/keeper.go

```diff
@@ -0,0 +1,289 @@
+		return fmt.Errorf("hold amounts %q for %s cannot be negative", funds, addr)
+	}
+
+	// Not bypassing hold's locked coins here because we're testing about new funds to be put on hold.
+	spendable := k.bankKeeper.SpendableCoins(ctx, addr)
+	for _, toAdd := range funds {
+		if toAdd.IsZero() {
+			continue
+		}
+		has, available := spendable.Find(toAdd.Denom)
+		if !has {
+			return fmt.Errorf("account %s spendable balance 0%s is less than hold amount %s", addr, toAdd.Denom, toAdd)
+		}
+		if available.Amount.LT(toAdd.Amount) {
+			return fmt.Errorf("account %s spendable balance %s is less than hold amount %s", addr, available, toAdd)
+		}
+	}
+
+	return nil
+}
+
+// AddHold puts the provided funds on hold for the provided account.
+func (k Keeper) AddHold(ctx sdk.Context, addr sdk.AccAddress, funds sdk.Coins, reason string) error {
+	if funds.IsZero() {
+		return nil
+	}
+
+	if err := k.ValidateNewHold(ctx, addr, funds); err != nil {
+		return err
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	var fundsAdded sdk.Coins
+	var errs []error
+	for _, toAdd := range funds {
+		if toAdd.IsZero() {
+			continue
+		}
+		onHold, err := k.getHoldCoinAmount(store, addr, toAdd.Denom)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to get current %s hold amount for %s: %w", toAdd.Denom, addr, err))
+			continue
+		}
+		newHoldAmt := onHold.Add(toAdd.Amount)
+		err = k.setHoldCoinAmount(store, addr, toAdd.Denom, newHoldAmt)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to place %s on hold for %s: %w", toAdd, addr, err))
+		}
+		fundsAdded = fundsAdded.Add(toAdd)
+	}
+
+	if !fundsAdded.IsZero() {
+		err := ctx.EventManager().EmitTypedEvent(hold.NewEventHoldAdded(addr, fundsAdded, reason))
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+// ReleaseHold releases the hold on the provided funds for the provided account.
+func (k Keeper) ReleaseHold(ctx sdk.Context, addr sdk.AccAddress, funds sdk.Coins) error {
+	if funds.IsZero() {
+		return nil
+	}
+	if funds.IsAnyNegative() {
+		return fmt.Errorf("cannot release %q from hold for %s: amounts cannot be negative", funds, addr)
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	var fundsReleased sdk.Coins
+	var errs []error
+	for _, toRelease := range funds {
+		if toRelease.IsZero() {
+			continue
+		}
+		onHold, err := k.getHoldCoinAmount(store, addr, toRelease.Denom)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to get current %s hold amount for %s: %w", toRelease.Denom, addr, err))
+			continue
+		}
+		newAmount := onHold.Sub(toRelease.Amount)
+		if newAmount.IsNegative() {
+			errs = append(errs, fmt.Errorf("cannot release %s from hold for %s: account only has %s%s on hold", toRelease, addr, onHold, toRelease.Denom))
+			continue
+		}
+		err = k.setHoldCoinAmount(store, addr, toRelease.Denom, newAmount)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to release %s from hold for %s: %w", toRelease, addr, err))
```

Comment:
I think you are missing a `continue` here. If `setHoldCoinAmount` fails then it will still get added to fundsReleased.

Reply:
Good catch. Fixed.